src/v/serde/json/parser.cc

```diff
@@ -0,0 +1,570 @@
+            vassert(
+              _frag_index != _frag_index_end,
+              "the check when we entered this function should prevent this");
+
+            size_t skip = std::min(
+              n, static_cast<size_t>(_frag_index_end - _frag_index));
+            _pos += skip;
+            n -= skip;
+
+            // NOLINTNEXTLINE(cppcoreguidelines-pro-bounds-pointer-arithmetic)
+            _frag_index += skip;
+            if (_frag_index == _frag_index_end) {
+                next_fragment();
+            }
+        }
+    }
+
+    ss::temporary_buffer<char> peek_buf() {
+        if (empty()) {
+            throw std::out_of_range("get on empty iobuf_wrapper");
+        }
+
+        return _frag->share(
+          _frag_index - _frag->get(), _frag_index_end - _frag_index);
+    }
+
+    bool empty() const { return _frag == _frag_end; }
+
+    // Return the number of bytes remaining in the buffer.
+    size_t remaining() const { return _size - _pos; }
+
+private:
+    void next_fragment() {
+        while (true) {
+            ++_frag;
+            if (_frag != _frag_end) {
+                _frag_index = _frag->get();
+                // NOLINTNEXTLINE(cppcoreguidelines-pro-bounds-pointer-arithmetic)
+                _frag_index_end = _frag->get() + _frag->size();
+                // handle an empty fragment
+                if (_frag_index == _frag_index_end) {
+                    continue;
+                }
+                return;
+            }
+            _frag_index = nullptr;
+            _frag_index_end = nullptr;
+            return;
+        }
+    }
+
+private:
+    iobuf _buf;
+    size_t _pos = 0;
+    size_t _size = 0;
+    iobuf::iterator _frag;
+    iobuf::const_iterator _frag_end;
+    const char* _frag_index = nullptr;
+    const char* _frag_index_end = nullptr;
+};
+
+class parser::impl {
+    using token = experimental::serde::json::token;
+
+public:
+    explicit impl(iobuf&& buf)
+      : _buf(std::move(buf)) {}
+
+    // The parser will suspend upon parsing a token that must be signaled to
+    // the caller.
+    ss::future<token> next_token() {
+        TRACE("next_token, remaining: {}\n", _buf.remaining());
+
+        // Reset the current value to a monostate to avoid reusing old
+        // values.
+        _current_value = std::monostate{};
+
+        co_await skip_whitespace();
+
+        // If suspension stack is empty, we are not expecting any more
+        // bytes in the buffer. However, if additional bytes are present
+        // it must mean that the input was malformed.
+        if (_suspension_stack.empty()) {
+            if (_buf.empty()) {
+                co_return suspend_with_token(token::eof);
+            } else {
+                co_return fuse_with_failure();
+            }
+        }
+
+        // Try to parse next byte(s) from the buffer based on the current
+        // suspension point.
+        switch (_suspension_stack.back()) {
+        case suspension_point::json_document:
```

Comment:
Does stuff top level values that are only leaf values (string, number, bool, null)?

Reply:
> The root element can be of any JSON data type. Note that in earlier RFC 4627, only objects or arrays were allowed as root values. Since the update to RFC 7159, the root of a JSON document can be of any JSON data type.

- [ ] Must be changed.